packages/client/src/objectSet/ObjectSetListenerWebsocket.ts

```diff
@@ -151,59 +217,87 @@ export class ObjectSetListenerWebsocket {
+    // we may have multiple subscriptions that don't have a subscriptionId yet,
+    // so we filter those out.
+    const readySubs = [...this.#subscriptions.values()].filter(isReady);
+    if (readySubs.length === 0) return;
+
+    const id = crypto.randomUUID();
+    // responses come back as an array of subIds, so we need to know the sources
+    this.#pendingSubscriptions.set(id, readySubs);
+
+    // every subscribe message "overwrites" the previous ones that are not
+    // re-included, we we have to reconsistute the entire list of subscriptions
+    const subscribe: ObjectSetSubscribeRequests = {
+      id,
+      requests: readySubs.map<ObjectSetSubscribeRequest>((
+        { temporaryObjectSetId },
+      ) => ({
+        objectSet: temporaryObjectSetId,
+        objectSetContext: {
+          objectSetFilterContext: { parameterOverrides: {} },
+        },
+        watchAllLinks: false,
+      })),
+    };
+
+    if (process.env.NODE_ENV !== "production") {
+      this.#logger?.debug(
+        { payload: subscribe },
+        "sending subscribe message",
+      );
     }
+
+    this.#ws?.send(JSON.stringify(subscribe));
   }
 
-  #expire(requestId: string) {
+  #expire(sub: Subscription) {
     // the temporary ObjectSet has expired, we should re-subscribe which will cause the
     // listener to get an onOutOfDate message when it becomes subscribed again
-    const state = this.#listeners.get(requestId);
-    if (state) {
-      const { subscriptionId, objectSet } = state;
-      if (subscriptionId) {
-        state.subscriptionId = undefined;
-        this.#subscriptionToRequestId.delete(subscriptionId);
-      }
-      this.#subscribe(requestId, objectSet);
-    }
+    sub.status = "expired";
+    this.#initiateSubscribe(sub);
   }
 
-  #unsubscribe(requestId: string) {
-    const data = this.#listeners.get(requestId);
-    if (data == null) {
-      return;
+  #unsubscribe(sub: Subscription) {
+    sub.status = "done";
+    // make sure listeners do nothing now
+    sub.listener = fillOutListener<any>({});
+    if (sub.expiry) {
+      clearTimeout(sub.expiry);
+      sub.expiry = undefined;
     }
-    this.#listeners.delete(requestId);
-    clearTimeout(data.expiry);
-    const { subscriptionId } = data;
-    if (subscriptionId != null) {
-      this.#subscriptionToRequestId.delete(subscriptionId);
-    }
-
-    if (this.#listeners.size === 0) {
-      this.#destroyWebsocket();
-    } else {
-      // TODO backend does not yet have an unsubscribe message payload
+    this.#subscriptions.delete(sub.subscriptionId);
+
+    // If we have no more subscriptions, we can disconnect the websocket
+    // however we should wait a bit to see if we get any more subscriptions.
+    // For example, when switching between react views, you may unsubscribe
+    // in the old view and subscribe in the new view. We don't need to re-establish
+    // the websocket connection in that case.
+    if (this.#maybeDisconnectTimeout) {
```

Comment:
if someone unsubscribed twice, is it bad that this timeout "resets" the subsequent times unsubscribe was called?

Reply:
Suppose you have one sub and remove it. 10s elapses. Then a new sub starts. After 4s we unsub. If we keep the original timer then after only 1s of no subs the connection is closed, instead of the 15s buffer I was trying to allow.

Reply:
Ah, nothing bad will happen. If the backend wants to protect against idle websocket connections they should handle that separetly. I'll guard too.